fetch-in-react-practice/client/components/app.jsx

```diff
@@ -0,0 +1,121 @@
+/* eslint-disable no-unused-vars -- Remove me */
+import React, { useEffect, useState } from 'react';
+import PageTitle from './page-title';
+import TodoList from './todo-list';
+import TodoForm from './todo-form';
+
+export default function App() {
+  const [todos, setTodos] = useState([]);
+
+  useEffect(() => {
+    /**
+     * Use fetch to send a GET request to `/api/todos`.
+     * Then 😉, once the response JSON is received and parsed,
+     * update state with the received todos.
+     *
+     */
+    fetch('api/todos')
+      .then((response) => response.json())
+      .then((data) => {
+        setTodos(data);
+      })
+      .catch((error) => {
+        console.error('There was an error!', error);
+      });
+  }, []);
+
+  function addTodo(newTodo) {
+    /**
+    * Use fetch to send a POST request to `/api/todos`.
+    * Then 😉, once the response JSON is received and parsed,
+    *   - create a shallow copy of the todos array from state
+    *   - add the todo received from the server to the copied array
+    *   - replace the old todos array in state with the new one
+    *
+    * DO NOT MUTATE the original state array, nor any objects within it.
+    * https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+    *
+    * TIP: Be sure to SERIALIZE the todo object in the body with JSON.stringify()
+    * and specify the "Content-Type" header as "application/json"
+    *
+    * TIP: Use Array.prototype.concat to create a new array containing the contents
+    * of the old array, plus the object returned by the server.
+    */
+
+    fetch('api/todos', {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json',
+      },
+      body: JSON.stringify(newTodo),
+    })
+      .then((response) => response.json())
+      .then((data) => {
+        const replica = [...todos];
+        const newTodos = replica.concat(data);
+        setTodos(newTodos);
+      })
+      .catch((error) => {
+        console.error('There was an error!', error);
+      });
+  }
+
+  function toggleCompleted(todoId) {
+    const index = todos.findIndex((todo) => todo.todoId === todoId);
+    const status = todos[index].isCompleted;
+    const completionStatus = { isCompleted: !status };
+
+    fetch(`/api/todos/${todoId}`, {
+      method: 'PATCH',
+      headers: {
+        'Content-Type': 'application/json',
+      },
+      body: JSON.stringify(completionStatus),
+    })
+      .then((response) => response.json())
+      .then((data) => {
+        const newTodos = [...todos];
+        newTodos.splice(index, 1, data);
+        setTodos(newTodos);
```

Comment:
Given the whole `PATCH` request could take a while, it's possible the patched todo item may not be in the same place as where it started.

Reply:
@junoulee Please re-calculate the index of your modified todo item within `todos`, _after_ the `PATCH` completes, but before you splice in the updated todo item from the server.